File: platform/api/build_identity.py
from __future__ import annotations

import os
import re


_FULL_GIT_SHA = re.compile(r"^[0-9a-f]{40}$")


def _clean(value: str | None, default: str) -> str:
    cleaned = (value or "").strip()
    if not cleaned or any(character in cleaned for character in "\r\n\x00"):
        return default
    return cleaned


def current_build_identity() -> dict[str, str]:
    revision = _clean(os.getenv("BMS_BUILD_SHA"), "unknown").lower()
    if not _FULL_GIT_SHA.fullmatch(revision):
        revision = "unknown"
    return {
        "revision": revision,
        "build_id": _clean(os.getenv("BMS_BUILD_ID"), "development"),
        "build_time": _clean(os.getenv("BMS_BUILD_TIME"), "unknown"),
    }

File: platform/desktop-electron/src/buildIdentity.ts
export type ElectronBuildIdentity = {
  layer: 'electron';
  revision: string;
  buildId: string;
  buildTime: string;
  appVersion: string;
};

const FULL_GIT_SHA = /^[0-9a-f]{40}$/;

function clean(value: string | undefined, fallback: string): string {
  return value?.trim() || fallback;
}

export function resolveElectronBuildIdentity(
  environment: NodeJS.ProcessEnv | Record<string, string | undefined>,
  appVersion: string,
): ElectronBuildIdentity {
  const rawRevision = clean(environment.BMS_BUILD_SHA, 'unknown');
  return Object.freeze({
    layer: 'electron',
    revision: FULL_GIT_SHA.test(rawRevision) ? rawRevision : 'unknown',
    buildId: clean(environment.BMS_BUILD_ID, 'development'),
    buildTime: clean(environment.BMS_BUILD_TIME, 'unknown'),
    appVersion: clean(appVersion, 'unknown'),
  });
}
